fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/PredicateCountByLevelProperty.java

```diff
@@ -0,0 +1,184 @@
+        }
+
+        /**
+         * Combine a list of PredicateCountByLevelInfo instances by adding the number of predicates at the same level
+         * across all instances.
+         *
+         * @param heightInfos a collection of {@code PredicateCountByLevelInfo} instances
+         * @return an instance of {@code PredicateCountByLevelInfo} with predicate counts added at each level.
+         */
+        public static PredicateCountByLevelInfo combine(Collection<PredicateCountByLevelInfo> heightInfos) {
+            final int highestLevel = heightInfos.stream().map(PredicateCountByLevelInfo::getHighestLevel).max(Integer::compare).orElse(0);
+            final var newLevelToComplexityMap = heightInfos
+                    .stream()
+                    .flatMap(heightInfo -> heightInfo.getLevelToPredicateCount().entrySet().stream())
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Integer::sum));
+            return new PredicateCountByLevelInfo(newLevelToComplexityMap, highestLevel);
+        }
+
+        /**
+         * Get a view of the predicate count at each level as a map.
+         *
+         * @return a map of level heights to the count of predicates at that level.
+         */
+        public Map<Integer, Integer> getLevelToPredicateCount() {
+            return Collections.unmodifiableMap(levelToPredicateCount);
+        }
+
+        /**
+         * Retrieves the highest level (i.e. height of root node) in this instance
+         *
+         * @return integer representing the highest level.
+         */
+        public int getHighestLevel() {
+            return highestLevel;
+        }
+
+        /**
+         * Compares two {@code PredicateCountByLevelInfo} instances level by level.
+         * This comparison is done by comparing the number of predicates at each level recorded within the two
+         * {@code PredicateCountByLevelInfo} instances, starting from the deepest level to the highest level,
+         * and returns the integer comparison between the first non-equal predicate counts.
+         *
+         * @param  a the first {@code PredicateCountByLevelInfo} to compare
+         * @param  b the second {@code PredicateCountByLevelInfo} to compare
+         * @return the value {@code 0} if {@code a} have the same number of predicates at each level as {@code b};
+         *         a value less than {@code 0} if a {@code a} has fewer predicates than {@code b} at a
+         *         deeper level or if {@code a} has a lower number of levels with predicates; and
+         *         a value greater than {@code 0} if {@code a} has more predicates than {@code b} at a
+         *         deeper level or if {@code a} has a higher number of levels with predicates.
+         */
+        public static int compare(final PredicateCountByLevelInfo a, final PredicateCountByLevelInfo b) {
+            final int highestLevel = Integer.max(a.getHighestLevel(), b.getHighestLevel());
```

Comment:
Wouldn't it be more concise to use two `SortedMap`s (`TreeMap`). You can then pop them off smallest first on both sides. In that way you can get rid of the `highestLevel` field which is redundant in your data structure.